File: app/src/main/java/mk/webfactory/template/util/NetworkUtils.java
package mk.webfactory.template.util;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import timber.log.Timber;

public class NetworkUtils {

  private final Context applicationContext;

  public NetworkUtils(Context context) {
    this.applicationContext = context.getApplicationContext();
  }

  public boolean isOnline() {
    return isOnline(applicationContext);
  }

  /**
   * Checks if there is an active internet connection on this device
   *
   * @return <code>true</code> if the device is connected to the internet;
   * <code>false</code> otherwise
   */
  public static boolean isOnline(Context context) {
    try {
      ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
      NetworkInfo netInfo = cm.getActiveNetworkInfo();
      return netInfo != null
          && netInfo.getState() == NetworkInfo.State.CONNECTED
          && (netInfo.getType() == ConnectivityManager.TYPE_WIFI
          || netInfo.getType() == ConnectivityManager.TYPE_MOBILE);
    } catch (Exception e) {
      Timber.e(e);
      return false;
    }
  }
}

File: app/src/main/java/mk/webfactory/template/util/NetworkUtils.kt
/*
 * MIT License
 *
 * Copyright (c) 2020 Web Factory LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package mk.webfactory.template.util

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.NetworkInfo
import android.os.Build
import timber.log.Timber

class NetworkUtils(context: Context) {
    private val applicationContext: Context = context.applicationContext
    val isOnline: Boolean
        get() = isOnline(applicationContext)

    companion object {
        /**
         * Checks if there is an active internet connection on this device
         *
         * @return true if the device is connected to the internet;
         * false otherwise
         */
        fun isOnline(context: Context): Boolean {
            return try {
                val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
                if (Build.VERSION.SDK_INT >= 29) {
                    val network = cm.activeNetwork
                    val networkCapabilities = cm.getNetworkCapabilities(network)
                    (network != null
                            && networkCapabilities != null
                            && networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                            && (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
                            || networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)))
                } else {
                    val netInfo = cm.activeNetworkInfo
                    netInfo != null
                            && netInfo.state == NetworkInfo.State.CONNECTED
                            && (netInfo.type == ConnectivityManager.TYPE_WIFI
                            || netInfo.type == ConnectivityManager.TYPE_MOBILE)
                }
            } catch (e: Exception) {
                Timber.e(e)
                false
            }
        }
    }

}